test/Kestrel.InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs

```diff
@@ -166,5 +169,114 @@ public async Task ResponseDrain_SlowerThanMinimumDataRate_AbortsConnection()
             _mockConnectionContext.Verify(c =>c.Abort(It.Is<ConnectionAbortedException>(e => 
                 e.Message == CoreStrings.ConnectionTimedBecauseResponseMininumDataRateNotSatisfied)), Times.Once);
         }
+
+        [Theory]
+        [InlineData(Http2FrameType.DATA)]
+        [InlineData(Http2FrameType.HEADERS)]
+        public async Task AbortedStream_ResetsAndDrainsRequest_RefusesFramesAfterCooldownExpires(Http2FrameType finalFrameType)
+        {
+            var mockSystemClock = _serviceContext.MockSystemClock;
+
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "POST"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+            };
+            await InitializeConnectionAsync(_appAbort);
+
+            await StartStreamAsync(1, headers, endStream: false);
+
+            await WaitForStreamErrorAsync(1, Http2ErrorCode.INTERNAL_ERROR, "The connection was aborted by the application.");
+
+            // There's a race when the appfunc is exiting about how soon it unregisters the stream.
+            for (var i = 0; i < 10; i++)
+            {
+                await SendDataAsync(1, new byte[100], endStream: false);
+            }
+
+            // Just short of the timeout
+            mockSystemClock.UtcNow += Constants.RequestBodyDrainTimeout;
+            (_connection as IRequestProcessor).Tick(mockSystemClock.UtcNow);
+
+            // Still fine
+            await SendDataAsync(1, new byte[100], endStream: false);
+
+            // Just past the timeout
+            mockSystemClock.UtcNow += TimeSpan.FromTicks(1);
+            (_connection as IRequestProcessor).Tick(mockSystemClock.UtcNow);
+
+            // Send an extra frame to make it fail
+            switch (finalFrameType)
+            {
+                case Http2FrameType.DATA:
+                    await SendDataAsync(1, new byte[100], endStream: true);
+                    break;
+
+                case Http2FrameType.HEADERS:
+                    await SendHeadersAsync(1, Http2HeadersFrameFlags.END_STREAM | Http2HeadersFrameFlags.END_HEADERS, _requestTrailers);
+                    break;
+
+                default:
+                    throw new NotImplementedException(finalFrameType.ToString());
+            }
+
+            await WaitForConnectionErrorAsync<Http2ConnectionErrorException>(ignoreNonGoAwayFrames: false, expectedLastStreamId: 1, Http2ErrorCode.STREAM_CLOSED,
+                CoreStrings.FormatHttp2ErrorStreamClosed(finalFrameType, 1));
+        }
+
+        [Fact]
+        public async Task DATA_Sent_TooSlowlyDueToSocketBackPressureOnSmallWrite_AbortsConnectionAfterGracePeriod()
```

Comment:
Were you planning to add any more tests?

Reply:
You need a flow control back pressure variant of this.

Reply:
Yep. I will be adding those.